### PR TITLE
reuse existing client

### LINKED ISSUE #15: Slow response when using S3
My response time increases from 60ms to 1500ms when using `RESULT_STORAGE = 'thumbor_aws.result_storage'` compared to no result storage. Any idea why or how to debug this?
I am coming from thumbor==6 with python2.7 and wanted to upgrade to version thumbor v7, also switching from tc_aws to thumbor-aws. Everything works fine except for this really high response time.

Installation
```
FROM python:3.10-slim
...
thumbor==7.0.8
thumbor-aws==0.4.0
```

thumbor.conf
```
RESULT_STORAGE = 'thumbor_aws.result_storage'
RESULT_STORAGE_EXPIRATION_SECONDS = 0
RESULT_STORAGE_STORES_UNSAFE = True

### AWS Storage ###
AWS_RESULT_STORAGE_REGION_NAME='eu-central-1'
AWS_RESULT_STORAGE_BUCKET_NAME='thumbor-adsadaaasd'
AWS_RESULT_STORAGE_ROOT_PATH='rs'
AWS_DEFAULT_LOCATION='https://thumbor-adsadaaasd.s3.eu-central-1.amazonaws.com'
```

I am using /.aws/config to assume aws role for authorization.

Edit: I think it is due to the fact that the boto aws session is not being reused. It looks a new session (via AWS assumeRole) is created for every single request.

### PR DESCRIPTION
**Pull request by guilhermef**
This PR is an attempt to reuse the same client across S3 requests, may fix #15

### CODE REVIEW

**Comment by heynemann** [comment]
Does it matter that we restart the client? From what I understood from the docs the client should be re-created every time (it does not hold any connections, does it?).

**Comment by guilhermef** [comment]
> Does it matter that we restart the client? From what I understood from the docs the client should be re-created every time (it does not hold any connections, does it?).

Based on this comment, it seems that the client can reuse the connections https://github.com/aio-libs/aiobotocore/issues/928

File: thumbor_aws/s3_client.py [hunk 1]
```diff
@@ -83,20 +83,21 @@ def file_acl(self) -> str:
...
+        """Singleton client for S3"""
+        if self.s3_client is None:
+            client = self.session.create_client(
+                "s3",
+                region_name=self.region_name,
+                aws_secret_access_key=self.secret_access_key,
+                aws_access_key_id=self.access_key_id,
+                endpoint_url=self.endpoint_url,
+            )
+            self.s3_client = await client.__aenter__()
```
Comment (heynemann) [inline thumbor_aws/s3_client.py:99]: We should be calling __ aexit__ shouldn't we? I think the correct way here is to assing self.s3_client to client and have the rest of the codebase do:

```python
with s3_client as client:
    # Whatever you need to do!
```

This way we ensure proper disposition of resources, right?

**Review by heynemann — CHANGES_REQUESTED** [review]

File: thumbor_aws/s3_client.py (hunk 1 above)
Comment (guilhermef) [reply thumbor_aws/s3_client.py:99]: We might not need all the changes I've made; the session object should be enough to reuse the same HTTP connections.

I would need @oliverschewe help to validate this.

Comment (guilhermef) [reply thumbor_aws/s3_client.py]: @heynemann I've updated the PR, to only reuse the S3Client object, but to still keep the `with s3_client as client` usage.

File: thumbor_aws/loader.py [hunk 2]
```diff
@@ -57,69 +58,84 @@
...
+                self._s3_client.configuration[
+                    "endpoint_url"
+                ] = context.config.TC_AWS_ENDPOINT
+                self._s3_client.configuration[
+                    "bucket_name"
+                ] = context.config.TC_AWS_LOADER_BUCKET
+                self._s3_client.configuration[
+                    "root_path"
+                ] = context.config.TC_AWS_LOADER_ROOT_PATH
+        return self._s3_client
```
Comment (heynemann) [inline thumbor_aws/loader.py]:
```suggestion
        if self._s3_client is not None:
                return self._s3_client

        self._s3_client = S3Client(context)
        self._s3_client.configuration = {
            "region_name": context.config.AWS_LOADER_REGION_NAME,
            "secret_access_key": context.config.AWS_LOADER_S3_SECRET_ACCESS_KEY,
            "access_key_id": context.config.AWS_LOADER_S3_ACCESS_KEY_ID,
            "endpoint_url": context.config.AWS_LOADER_S3_ENDPOINT_URL,
            "bucket_name": context.config.AWS_LOADER_BUCKET_NAME,
            "root_path": context.config.AWS_LOADER_ROOT_PATH,
        }

        if self._s3_client.compatibility_mode is True:
            self._s3_client.configuration[
                "region_name"
            ] = context.config.TC_AWS_REGION
            self._s3_client.configuration[
                "endpoint_url"
            ] = context.config.TC_AWS_ENDPOINT
            self._s3_client.configuration[
                "bucket_name"
            ] = context.config.TC_AWS_LOADER_BUCKET
            self._s3_client.configuration[
                "root_path"
            ] = context.config.TC_AWS_LOADER_ROOT_PATH

        return self._s3_client
```

File: thumbor_aws/loader.py [hunk 3]
```diff
@@ -57,69 +58,84 @@
...
-
-    if status_code != 200:
-        result.error = LoaderResult.ERROR_NOT_FOUND
-        result.extra = body
-        result.successful = False
-        return result
-
-    result.successful = True
-    result.buffer = body
+class S3Loader:
```
Comment (heynemann) [inline thumbor_aws/loader.py]: I get why you added a class to make it a singleton but we can make a singleton just as easily with methods:

```python
__instance = None

def load(params):
    if __instance is None:
        __instance = Whatever()

    # do something with __instance
```

**Review by heynemann — APPROVED** [review]
LGTM with minor nits!

**Review by heynemann — APPROVED** [review]
Much clearer! TYVM!

**Pull request merged**